rust/ql/lib/codeql/rust/internal/TypeInference.qll

```diff
@@ -1178,6 +1198,120 @@ private Type inferIndexExprType(IndexExpr ie, TypePath path) {
   )
 }
 
+/**
+ * A matching configuration for resolving types of struct patterns
+ * like `let Foo { bar } = ...`.
+ */
+private module StructPatMatchingInput implements MatchingInputSig {
+  class DeclarationPosition = StructExprMatchingInput::DeclarationPosition;
+
+  class Declaration = StructExprMatchingInput::Declaration;
+
+  class AccessPosition = DeclarationPosition;
+
+  class Access extends StructPat {
+    Type getTypeArgument(TypeArgumentPosition apos, TypePath path) { none() }
```

Comment:
Why is this `none()`?

Reply:
Since https://github.com/github/codeql/pull/19847, `getTypeArgument` is only used for type arguments supplied to functions, not to their defining type, i.e. it is used for `foo::<i32>(...)` but not for `Bar::<i32>::foo(...)`.

Since explicit type arguments in pattern matching corresponds to the latter, it is `none()` here, and instead handled in `getInferredType` below. I'll add some tests for this.

Reply:
Adding test, of course, revealed some corner cases that were not handled, I'm gonna do that on this PR as well.